### PR TITLE
Updated Blame client side code for clarity

### PR DESCRIPTION
**Pull request by KieranRatcliffeInvertedAI**
Summary of changes:
-Updated documentation of blame() and BlameResponse for clarity
-BlameResponse.blamed_result=>BlameResponse.agents_at_fault
-blame(...,candidate_agents)=>blame(...,colliding_agents)

test_blame.py and mock.py were updated to reflect the changes in these names.

### CODE REVIEW

**Comment by zhangk1551** [comment]
Could you update the affected examples as well?
[examples/blame_example.ipynb](https://github.com/inverted-ai/invertedai/blob/5bd962be12ab75de6dd4ba0b14a2740252d027cd/examples/blame_example.ipynb)
[examples/blame_minimal_example.py](https://github.com/inverted-ai/invertedai/pull/128/files#diff-1e9e58a4d5ac94c83a71055774169569fd6bdd36d6d053423985c0f254153308
)

**Comment by KieranRatcliffeInvertedAI** [comment]
> Could you update the affected examples as well? [examples/blame_example.ipynb](https://github.com/inverted-ai/invertedai/blob/5bd962be12ab75de6dd4ba0b14a2740252d027cd/examples/blame_example.ipynb) [examples/blame_minimal_example.py](https://github.com/inverted-ai/invertedai/pull/128/files#diff-1e9e58a4d5ac94c83a71055774169569fd6bdd36d6d053423985c0f254153308)

Yes I can make those changes.

**Comment by KieranRatcliffeInvertedAI** [comment]
@zhangk1551 For the blame_minimal_example.py demo, it seems to rely upon the detect_collisions feature which is going to be removed. My thoughts are we simply delete this file until I finish the official blame demos to replace it. What are your thoughts?

**Comment by zhangk1551** [comment]
@KieranRatcliffeInvertedAI That's fine. Could you provide at least one example (no matter it's new or modified from the old one), that I can use to test the new API after I update the server side?

**Comment by KieranRatcliffeInvertedAI** [comment]
> @KieranRatcliffeInvertedAI That's fine. Could you provide at least one example (no matter it's new or modified from the old one), that I can use to test the new API after I update the server side?

I modified the example notebook so that you can test changes on the server side.

**Comment by Ruishenl** [comment]
@zhangk1551  I will leave the approval of this pr to you then. Once merged i will make a new dev-release from the latest change

**Comment by KieranRatcliffeInvertedAI** [comment]
I made another change to the blame agent_state_history documentation that I hope clarifies the intended use of Blame to the user. I would like feedback on the documentation since we don't have hard guidelines on some aspects of the agent_state_history. Please review the whole files documentation but here is the excerpt of interest: 

"
        Lists containing AgentState objects for every agent within the scene (up to 100
        agents) for each time step within the relevant sequence immediately preceding 
        the collision.
        The final list of AgentState objects should include the first time step of the 
        collision and no time steps afterwards. The lists of AgentState objects preceding 
        the collision should capture enough of the scenario context before the collision 
        for the Blame model to analyze and assign fault. 30-100 time steps preceding the 
        the collision is recommended for best results.
        Each AgentState state must include x: [float], y: [float] corrdinate in meters 
        orientation: [float] in radians with 0 pointing along x and pi/2 pointing along 
        y and speed: [float] in m/s.
"

File: invertedai/api/blame.py
Comment (Ruishenl) [inline invertedai/api/blame.py:65]: typo ?

File: invertedai/api/blame.py
Comment (Ruishenl) [inline invertedai/api/blame.py:66]: I would add  it's  positive x-axis and positive y-axis

File: invertedai/api/blame.py
Comment (KieranRatcliffeInvertedAI) [reply invertedai/api/blame.py:65]: Yes thank you.

File: invertedai/api/blame.py
Comment (KieranRatcliffeInvertedAI) [reply invertedai/api/blame.py:66]: Done.

**Comment by zhangk1551** [comment]
> 

Hi, @KieranRatcliffeInvertedAI , could you emphasize the gap of the two adjacent time step is 0.1 second? And the history for the time steps are supposed to be continuous (i.e. no missing time step  in the middle).



> @zhangk1551 I will leave the approval of this pr to you then. Once merged i will make a new dev-release from the latest change

Nice. I am modifying the code on the server side. I took a glance at the code in this PR, and it looked good overall. Let's hold this PR for a little while, and merge it together with the corresponding server change.

**Comment by KieranRatcliffeInvertedAI** [comment]
Here's the updated documentation for agent_state_history:

"
agent_state_history:
        Lists containing AgentState objects for every agent within the scene (up to 100
        agents) for each time step within the relevant continuous sequence immediately 
        preceding the collision.
        The list of AgentState objects should include the first time step of the collision
        and no time steps afterwards. The lists of AgentState objects preceding the 
        collision should capture enough of the scenario context before the collision for 
        the Blame model to analyze and assign fault. For best results it is recommended to 
        input 30-100 time steps of 0.1s each preceding the collision.
        Each AgentState state must include x: [float], y: [float] coordinates in meters, 
        orientation: [float] in radians with 0 pointing along the positive x axis and 
        pi/2 pointing along the positive y axis, and speed: [float] in m/s.
"

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:126]: I don't think the code logic here matches the colliding agent pairs correctly, especially when there are more than one collisions in the same frame. 

You can copy the [function](https://github.com/inverted-ai/driving-models/blob/5559f4c53580a27692bdbec7c2cd2769b63eeeca/api_server/app/utils.py#L448) in the server side and change accordingly based on your needs. The idea is to match the colliding agent pair based on the overlap.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:133]: It would be better not to put them in the same line.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:108]: `t in range(100)` is not 20 seconds.

Besides, since you don't use the data after the first collision, you can stop immediately as soon as `is_found_collision`.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:147]: If I understand correctly, you are supposed to use `collisions` directly as `colliding_agents` here.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:150]: Didn't replace `candidate_agents` to `colliding_agents`.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:158]: ditto

**Review by zhangk1551 — CHANGES_REQUESTED** [review]
@KieranRatcliffeInvertedAI 

Sorry that I didn't read the `example` carefully before. I don't think the notebook works. 

This PR contain two parts:
1. Update the client side so that it won't depend on the collision detection on the server side.
2. Adjust the variable names.

As a good practice, could you do item 1 first, test it with the current server API, and ensure it's issue-free?
And then please replace the variable names (item 2), and check them manually, which you won't be able to test without the server updates.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:126]: That can be done, this will be temporary until we create a collision detection feature within the invertedai library and should be replaced then.

Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:126]: Now that I've looked into this a bit more, I am realizing the complications. This repo is public facing but the function you linked is from a private repo thus cannot be included. We do not have a public facing collision detection feature except the one in drive that is limited. I think the solution is that there are a selection of logs that can be played instead of generating a collision scenario live. What are your thoughts on this @zhangk1551 since it will change this example quite a lot?

Comment (zhangk1551) [reply examples/blame_example.ipynb:126]: For this particular [function](https://github.com/inverted-ai/driving-models/blob/5559f4c53580a27692bdbec7c2cd2769b63eeeca/api_server/app/utils.py#L448), it's standalone and doesn't rely on other functions in the private repo. We can move it to the public client repo directly.

Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: A collision detection function has been added to the demo notebook.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Fixed.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Fixed.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Fixed.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Fixed.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Fixed.

**Review by zhangk1551 — COMMENTED** [review]
For the demo, could you attach gifs of collision cases and the corresponding collision detection results and blamed results here?

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:186]: What if a pair of agents have two collisions in the scenario? i.e. They collide, and separate, and collide again?

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:98]: Why do you prefer writing a new function here instead of using the existing function from the `driving_model` repo? It feels like reinventing the wheel, and make the code base hard to manage.

Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:98]: That particular function appears to made for performing collision detection in a specific, inapplicable scenario that requires pre-processed collision information as input that is not available in this demo notebook. The function I have provided is more general and takes our standard, common IAI data types as input (e.g. AgentState and AgentAttributes) making it more accessible for clients and other users. Furthermore it is a bit lighter-weight and more transparent how the collision is detected. 

I do not think we will be providing this as a tool for general use so this function would only exist in this notebook anyways and we would furthermore not be importing from driving_models since this notebook needs to work for clients who won't have access to that repository.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:186]: I have added a fix to that so the same collision pair can be added multiple times.

**Review by zhangk1551 — CHANGES_REQUESTED** [review]

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:180]: Yes python throws an error when you try to remove a key-value pair from a dictionary while iterating over it, resulting in ugly code like this.

Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:180]: I've updated the secondary loop to be a bit cleaner.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:135]: I'm not totally sure what this question means but if you're asking why I used a numpy array instead of a list, even though matmul() can operate without numpy arrays, this is done for cleanliness and future-proofing.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:110]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:145]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:175]: I changed this data structure to deal only with tuples

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:156]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:174]: This is indeed superfluous logic and has been removed.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:178]: Good catch, that logic has been updated so that they are unified.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:135]: Oh I see, I put the cosine in the wrong element. This has been corrected.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:139]: I transposed the R, V, RV (now VR) and polygons matrices so that:
`polygons = [Polygon([p[:,0], p[:,1], p[:,2], p[:,3]]) for p in polygons]\n"`
is now:
`polygons = [Polygon([p[0], p[1], p[2], p[3]]) for p in polygons]`

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:155]: Done

**Comment by KieranRatcliffeInvertedAI** [comment]
> For the demo, could you attach gifs of collision cases and the corresponding collision detection results and blamed results here?

These GIFs were captured using the original parameter names since those changes have not been pushed to the server yet.

[LogCollision(collision_agents=(4, 14), start_time=4, end_time=7)]
agents_at_fault: (4,)
reasons: {4: ['rear_end']}
confidence_score: 0.6062055826187134
blame-example

[LogCollision(collision_agents=(4, 12), start_time=22, end_time=65)]
agents_at_fault: (12,)
reasons: {12: ['traffic_light_violation', 'rear_end']}
confidence_score: 0.6315696835517883
blame-example

[LogCollision(collision_agents=(1, 2), start_time=48, end_time=99)]
agents_at_fault: (2,)
reasons: {2: ['rear_end']}
confidence_score: 0.9370070695877075
blame-example

[LogCollision(collision_agents=(5, 8), start_time=25, end_time=36), LogCollision(collision_agents=(2, 11), start_time=57, end_time=99)]
agents_at_fault: (8,), (11,)
reasons: {8: ['rear_end']}, {11: ['rear_end']}
confidence_score:0.9205628633499146, 0.3135790228843689
blame-example

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:139]: How is it different from `polygons = [Polygon(p) for p in polygons]`?

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:93]: It would be nice if you can wrap some code blocks in this function into new functions and try using meaningful variable/function names. In this way, the code can be self-explanatory and we don't need too many inline comments.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:171]:
```
        collisions_all.append(collisions_ongoing[agent_tuple])
        keys_to_delete.append(agent_tuple)
```
can be shared here.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:178]: Averagely, how long does it take to generate a collision?

**Review by zhangk1551 — COMMENTED** [review]
Curious that why there are black lines near the traffic lights.

**Comment by Ruishenl** [comment]
> Curious that why there are black lines near the traffic lights.

Black line are rendered in static image on server side when no states are passed. The client image is based on top of that

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:178]: I haven't captured any data in that regards but anecdotally I've never see it go beyond 3 attempts. I just didn't want to leave the possibility of an infinite loop if the end user plays around with some settings that makes collisions unlikely.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:139]: According to the Polygon documentation, this would be the same so I've changed it to be the concise option you've put here.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:138]: Several variable names were changed to be clearer including this one.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:93]: Any code blocks that do not reference above-scope variables have been placed into their own functions for clarity.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:171]: Done.

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:157]: `len(agent_states) != len(agent_attributes)`

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:164]: `if agent_tuple not in collisions_ongoing`

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:172]: `agent_tuple not in ...`

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:179]: `collisions_ongoing[agent_tuple].end_time is not None`

**Review by zhangk1551 — COMMENTED** [review]
The code looks much nicer. Some minor comments about `not`.

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:157]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:164]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:172]: Done

File: examples/blame_example.ipynb
Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb:179]: Done

File: examples/blame_example.ipynb
Comment (zhangk1551) [inline examples/blame_example.ipynb:180]: `agent_tuple not in`

be careful...

Comment (KieranRatcliffeInvertedAI) [reply examples/blame_example.ipynb]: Done

**Review by zhangk1551 — APPROVED** [review]
LGTM.

Let's ping @adscib @Ruishenl to see if they have other concerns.

**Comment by Ruishenl** [comment]
Feel free to merge it.  The server side still needs a pr so i can release it if you guys are ready

**Comment by KieranRatcliffeInvertedAI** [comment]
> Feel free to merge it. The server side still needs a pr so i can release it if you guys are ready

Please open that PR as soon as this one is merged.

**Comment by KieranRatcliffeInvertedAI** [comment]
For documentation purposes, I received verbal confirmation from Adam to go ahead and merge this PR.

**Pull request merged**
